tmt/steps/report/junit.py

```diff
@@ -83,24 +82,36 @@ class PropertyDict(TypedDict):
         name: str
         value: str
 
-    def __init__(self, wrapped: tmt.Result) -> None:
-        self._wrapped = wrapped
+    def __init__(self) -> None:
         self._properties: dict[str, str] = {}
 
     @property
     def properties(self) -> list[PropertyDict]:
         return [{'name': k, 'value': v} for k, v in self._properties.items()]
 
     @properties.setter
-    def properties(self, keyval: dict[str, str]) -> None:
-        self._properties = keyval
+    def properties(self, keyval: dict[str, Optional[str]]) -> None:
```

Comment:
@happz If you remember we discussed incompatible mypy types if I use `@properties.setter` instead of defining a separate setter method (e.g. `set_properties()`).

The mypy doesn't like that the property setter accepts the `dict` and a property getter returns a `list` of properties.

The error is:
```
tmt/steps/report/polarion.py:313: error: Incompatible types in assignment (expression has type "dict[str, str]", variable has type "list[PropertyDict]")  [assignment]
tmt/steps/report/polarion.py:323: error: Incompatible types in assignment (expression has type "dict[str, Any]", variable has type "list[PropertyDict]")  [assignment]
```

Do you have any hints/recommendations on how to solve it?

Reply:
Ok, so there's an issue for mypy discussing this (started with "it's a perversion we don't want to support" :), so I ended up with this:

```diff
diff --git a/tmt/steps/report/polarion.py b/tmt/steps/report/polarion.py
index 1f7c4d53..5f92dbd3 100644
--- a/tmt/steps/report/polarion.py
+++ b/tmt/steps/report/polarion.py
@@ -259,7 +259,7 @@ class ReportPolarion(tmt.steps.report.ReportPlugin[ReportPolarionData]):
             'pool_team',
             'sample_image']
 
-        testsuites_properties = {}
+        testsuites_properties: dict[str, Optional[str]] = {}
 
         for tr_field in other_testrun_fields:
             param = self.get(tr_field, os.getenv(f'TMT_PLUGIN_REPORT_POLARION_{tr_field.upper()}'))
@@ -310,7 +310,10 @@ class ReportPolarion(tmt.steps.report.ReportPlugin[ReportPolarionData]):
                 'polarion-testcase-project-id': test_project_id,
                 }
 
-            result.properties = testcase_properties
+            # ignore[assignment]: mypy does not support different types for property getter
+            # and setter. The assignment is correct, but mypy cannot tell.
+            # See https://github.com/python/mypy/issues/3004 for getter/setter discussions
+            result.properties = testcase_properties  # type: ignore[assignment]
 
         assert self.workdir is not None
 
@@ -320,7 +323,10 @@ class ReportPolarion(tmt.steps.report.ReportPlugin[ReportPolarionData]):
             'polarion-testrun-title': title,
             'polarion-project-span-ids': ','.join([project_id, *project_span_ids])})
 
-        results_context.properties = testsuites_properties
+        # ignore[assignment]: mypy does not support different types for property getter
+        # and setter. The assignment is correct, but mypy cannot tell.
+        # See https://github.com/python/mypy/issues/3004 for getter/setter discussions
+        results_context.properties = testsuites_properties  # type: ignore[assignment]
 
         xml_data = make_junit_xml(
             phase=self,
```

Reply:
:D, ok, thanks. I did the modifications, marking this as resolved.